File: harness/parity/go/transaction.go
package main

import (
	"strings"

	memless "github.com/nascent-tech/memless-go"
)

func transactionOutcome(fixture string, suite string) string {
	dir, dest, base := copyFixture(fixture)
	defer cleanupDir(dir)
	instance, failed := loadInstance(dest)
	if instance == nil {
		return failed
	}
	return reportSuite(instance, dir, dest, base, suite)
}

func transactionDiskOutcome(fixture string, suite string) string {
	dir, dest, base := copyFixture(fixture)
	defer cleanupDir(dir)
	instance, failed := loadInstance(dest)
	if instance == nil {
		return failed
	}
	return withReadonlyDir(dir, func() string { return reportSuite(instance, dir, dest, base, suite) })
}

func reportSuite(instance *memless.Instance, dir string, dest string, base string, suite string) string {
	defer instance.Release()
	var out strings.Builder
	for _, sql := range strings.Split(suite, ";;") {
		out.WriteString(runInstruction(instance, sql, dest, base))
		out.WriteString("\n")
	}
	out.WriteString("sha256:" + fileHash(dest) + "\nresidue:" + residueFlag(dir, base) + "\n")
	return out.String()
}

func runInstruction(instance *memless.Instance, sql string, dest string, base string) string {
	if isSelect(sql) {
		return queryInstruction(instance, sql)
	}
	return runWrite(instance, sql, dest, base)
}

func isSelect(sql string) bool {
	return strings.HasPrefix(strings.ToUpper(strings.TrimSpace(sql)), "SELECT")
}

func queryInstruction(instance *memless.Instance, sql string) string {
	rows, err := instance.Query(sql)
	if err != nil {
		return errorLine(err)
	}
	return strings.TrimRight(render(rows), "\n")
}

File: harness/parity/php/transaction.php
<?php

declare(strict_types=1);

use Memless\Instance;
use Memless\MemlessRefusal;

function transactionOutcome(string $fixture, string $suite): string
{
    $paths = copyFixture($fixture);
    [$instance, $failed] = loadOrFail($paths[1]);
    $report = $instance !== null ? reportSuite($instance, $suite, $paths) : $failed;
    cleanupDir($paths[0]);

    return $report;
}

function transactionDiskOutcome(string $fixture, string $suite): string
{
    $paths = copyFixture($fixture);
    [$instance, $failed] = loadOrFail($paths[1]);
    $report = $instance !== null
        ? withReadonlyDir($paths[0], fn () => reportSuite($instance, $suite, $paths))
        : $failed;
    cleanupDir($paths[0]);

    return $report;
}

function reportSuite(Instance $instance, string $suite, array $paths): string
{
    [$dir, $dest, $base] = $paths;
    $out = '';
    foreach (explode(';;', $suite) as $sql) {
        $out .= runInstruction($instance, $sql, $dest, $base) . "\n";
    }
    $out .= 'sha256:' . fileHash($dest) . "\nresidue:" . residueFlag($dir, $base) . "\n";
    $instance->release();

    return $out;
}

function runInstruction(Instance $instance, string $sql, string $dest, string $base): string
{
    if (str_starts_with(strtoupper(trim($sql)), 'SELECT')) {
        return queryInstruction($instance, $sql);
    }

    return runWrite($instance, $sql, $dest, $base);
}

function queryInstruction(Instance $instance, string $sql): string
{
    try {
        return rtrim(render($instance->query($sql)), "\n");
    } catch (MemlessRefusal $refusal) {
        return 'refused: ' . $refusal->getMessage();
    } catch (\Throwable $fault) {
        return 'fault: ' . $fault->getMessage();
    }
}
